File: hypersim-sdk/src/security/SecurityAuditor.ts
/**
 * Security audit logging and monitoring
 */

import { EventEmitter } from 'events';
import fs from 'fs/promises';
import path from 'path';
import os from 'os';
import { SecurityEvent, SecurityEventType } from './types';

interface AuditConfig {
  logDirectory?: string;
  maxLogSize?: number;
  rotateDaily?: boolean;
  enableRemoteLogging?: boolean;
  remoteEndpoint?: string;
  bufferSize?: number;
  flushInterval?: number;
}

export class SecurityAuditor extends EventEmitter {
  private readonly config: Required<AuditConfig>;
  private readonly eventBuffer: SecurityEvent[] = [];
  private flushTimer: NodeJS.Timeout;
  private currentLogFile?: string;
  private logRotationTimer: NodeJS.Timeout;

  constructor(enabled: boolean, config: AuditConfig = {}) {
    super();
    
    this.config = {
      logDirectory: config.logDirectory ?? path.join(os.tmpdir(), '.hypersim-audit'),
      maxLogSize: config.maxLogSize ?? 10 * 1024 * 1024, // 10MB
      rotateDaily: config.rotateDaily ?? true,
      enableRemoteLogging: config.enableRemoteLogging ?? false,
      remoteEndpoint: config.remoteEndpoint ?? '',
      bufferSize: config.bufferSize ?? 100,
      flushInterval: config.flushInterval ?? 30000 // 30 seconds
    };

    if (enabled) {
      this.initialize();
    }
  }

  /**
   * Log security event
   */
  async log(event: SecurityEvent): Promise<void> {
    // Add to buffer
    this.eventBuffer.push({
      ...event,
      timestamp: event.timestamp || Date.now()
    });

    // Emit event for real-time monitoring
    this.emit('securityEvent', event);

    // Flush if buffer is full or event is critical
    if (this.eventBuffer.length >= this.config.bufferSize || event.severity === 'critical') {
      await this.flush();
    }
  }

  /**
   * Log authentication event
   */
  async logAuth(success: boolean, userId?: string, ip?: string, userAgent?: string): Promise<void> {
    await this.log({
      type: success ? SecurityEventType.UNAUTHORIZED_ACCESS : SecurityEventType.UNAUTHORIZED_ACCESS,
      severity: success ? 'low' : 'medium',
      timestamp: Date.now(),
      description: success ? 'Successful authentication' : 'Failed authentication attempt',
      metadata: {
        userId,
        ip,
        userAgent,
        success
      }
    });
  }

  /**
   * Log API usage
   */
  async logAPIUsage(endpoint: string, method: string, statusCode: number, responseTime: number, userId?: string): Promise<void> {
    const severity = statusCode >= 400 ? 'medium' : 'low';
    
    await this.log({
      type: SecurityEventType.UNAUTHORIZED_ACCESS,
      severity,
      timestamp: Date.now(),
      description: `API ${method} ${endpoint} - ${statusCode}`,
      metadata: {
        endpoint,
        method,
        statusCode,
        responseTime,
        userId
      }
    });
  }

  /**
   * Log data access
   */
  async logDataAccess(resource: string, action: 'read' | 'write' | 'delete', userId?: string, success: boolean = true): Promise<void> {
    await this.log({
      type: SecurityEventType.UNAUTHORIZED_ACCESS,
      severity: success ? 'low' : 'high',
      timestamp: Date.now(),
      description: `Data ${action} on ${resource}`,
      metadata: {
        resource,
        action,
        userId,
        success
      }
    });
  }

  /**
   * Flush buffered events to storage
   */
  async flush(): Promise<void> {
    if (this.eventBuffer.length === 0) {
      return;
    }

    const events = [...this.eventBuffer];
    this.eventBuffer.length = 0; // Clear buffer

    try {
      // Write to file
      await this.writeToFile(events);

      // Send to remote endpoint if enabled
      if (this.config.enableRemoteLogging && this.config.remoteEndpoint) {
        await this.sendToRemote(events);
      }

      this.emit('eventsFlushed', { count: events.length });
    } catch (error) {
      // Add events back to buffer if write failed
      this.eventBuffer.unshift(...events);
      this.emit('flushError', error);
    }
  }

  /**
   * Query audit logs
   */
  async queryLogs(filter: {
    startDate?: Date;
    endDate?: Date;
    severity?: string;
    type?: SecurityEventType;
    userId?: string;
    limit?: number;
  }): Promise<SecurityEvent[]> {
    const logFiles = await this.getLogFiles();
    const results: SecurityEvent[] = [];

    for (const file of logFiles) {
      try {
        const content = await fs.readFile(file, 'utf8');
        const lines = content.split('\n').filter(line => line.trim());
        
        for (const line of lines) {
          try {
            const event = JSON.parse(line);
            if (this.matchesFilter(event, filter)) {
              results.push(event);
              
              if (filter.limit && results.length >= filter.limit) {
                return results;
              }
            }
          } catch {
            // Skip invalid lines
          }
        }
      } catch {
        // Skip unreadable files
      }
    }

    return results.sort((a, b) => b.timestamp - a.timestamp);
  }

  /**
   * Get security statistics
   */
  async getStats(timeRange: { start: Date; end: Date }): Promise<{
    totalEvents: number;
    eventsBySeverity: Record<string, number>;
    eventsByType: Record<string, number>;
    topIPs: Array<{ ip: string; count: number }>;
    topUsers: Array<{ userId: string; count: number }>;
  }> {
    const events = await this.queryLogs({
      startDate: timeRange.start,
      endDate: timeRange.end
    });

    const stats = {
      totalEvents: events.length,
      eventsBySeverity: {} as Record<string, number>,
      eventsByType: {} as Record<string, number>,
      topIPs: [] as Array<{ ip: string; count: number }>,
      topUsers: [] as Array<{ userId: string; count: number }>
    };

    const ipCounts = new Map<string, number>();
    const userCounts = new Map<string, number>();

    for (const event of events) {
      // Count by severity
      stats.eventsBySeverity[event.severity] = (stats.eventsBySeverity[event.severity] || 0) + 1;
      
      // Count by type
      stats.eventsByType[event.type] = (stats.eventsByType[event.type] || 0) + 1;
      
      // Count IPs
      const ip = event.metadata?.ip;
      if (ip) {
        ipCounts.set(ip, (ipCounts.get(ip) || 0) + 1);
      }
      
      // Count users
      const userId = event.metadata?.userId;
      if (userId) {
        userCounts.set(userId, (userCounts.get(userId) || 0) + 1);
      }
    }

    // Top IPs
    stats.topIPs = Array.from(ipCounts.entries())
      .map(([ip, count]) => ({ ip, count }))
      .sort((a, b) => b.count - a.count)
      .slice(0, 10);

    // Top users
    stats.topUsers = Array.from(userCounts.entries())
      .map(([userId, count]) => ({ userId, count }))
      .sort((a, b) => b.count - a.count)
      .slice(0, 10);

    return stats;
  }

  /**
   * Archive old logs
   */
  async archiveLogs(olderThanDays: number): Promise<void> {
    const cutoffDate = new Date(Date.now() - olderThanDays * 24 * 60 * 60 * 1000);
    const logFiles = await this.getLogFiles();

    for (const file of logFiles) {
      try {
        const stats = await fs.stat(file);
        if (stats.mtime < cutoffDate) {
          const archivePath = file.replace('.log', '.archive');
          await fs.rename(file, archivePath);
          this.emit('logArchived', { file, archivePath });
        }
      } catch (error) {
        this.emit('archiveError', { file, error });
      }
    }
  }

  /**
   * Shutdown auditor
   */
  async shutdown(): Promise<void> {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
    }
    
    if (this.logRotationTimer) {
      clearInterval(this.logRotationTimer);
    }

    await this.flush();
  }

  private async initialize(): Promise<void> {
    // Create log directory
    try {
      await fs.mkdir(this.config.logDirectory, { recursive: true, mode: 0o700 });
    } catch (error) {
      console.error('Failed to create audit log directory:', error);
    }

    // Set up periodic flushing
    this.flushTimer = setInterval(async () => {
      await this.flush();
    }, this.config.flushInterval);

    // Set up daily log rotation if enabled
    if (this.config.rotateDaily) {
      this.setupLogRotation();
    }

    this.currentLogFile = this.getCurrentLogFile();
  }

  private async writeToFile(events: SecurityEvent[]): Promise<void> {
    if (!this.currentLogFile) {
      this.currentLogFile = this.getCurrentLogFile();
    }

    const logEntries = events.map(event => JSON.stringify(event)).join('\n') + '\n';
    
    try {
      await fs.appendFile(this.currentLogFile, logEntries, { mode: 0o600 });
      
      // Check file size for rotation
      const stats = await fs.stat(this.currentLogFile);
      if (stats.size > this.config.maxLogSize) {
        await this.rotateLog();
      }
    } catch (error) {
      console.error('Failed to write audit log:', error);
      throw error;
    }
  }

  private async sendToRemote(events: SecurityEvent[]): Promise<void> {
    try {
      const response = await fetch(this.config.remoteEndpoint, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'User-Agent': 'HyperSim-SDK-Auditor'
        },
        body: JSON.stringify({ events })
      });

      if (!response.ok) {
        throw new Error(`Remote logging failed: ${response.status}`);
      }
    } catch (error) {
      this.emit('remoteLoggingError', error);
      throw error;
    }
  }

  private getCurrentLogFile(): string {
    const today = new Date().toISOString().split('T')[0];
    return path.join(this.config.logDirectory, `security-audit-${today}.log`);
  }

  private async rotateLog(): Promise<void> {
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const rotatedFile = this.currentLogFile!.replace('.log', `-${timestamp}.log`);
    
    try {
      await fs.rename(this.currentLogFile!, rotatedFile);
      this.currentLogFile = this.getCurrentLogFile();
      this.emit('logRotated', { oldFile: rotatedFile, newFile: this.currentLogFile });
    } catch (error) {
      this.emit('rotationError', error);
    }
  }

  private setupLogRotation(): void {
    // Rotate at midnight
    const now = new Date();
    const tomorrow = new Date(now);
    tomorrow.setDate(tomorrow.getDate() + 1);
    tomorrow.setHours(0, 0, 0, 0);
    
    const msUntilMidnight = tomorrow.getTime() - now.getTime();
    
    setTimeout(() => {
      this.rotateLog();
      
      // Set up daily rotation
      this.logRotationTimer = setInterval(() => {
        this.rotateLog();
      }, 24 * 60 * 60 * 1000);
    }, msUntilMidnight);
  }

  private async getLogFiles(): Promise<string[]> {
    try {
      const files = await fs.readdir(this.config.logDirectory);
      return files
        .filter(file => file.endsWith('.log'))
        .map(file => path.join(this.config.logDirectory, file))
        .sort();
    } catch {
      return [];
    }
  }

  private matchesFilter(event: SecurityEvent, filter: any): boolean {
    if (filter.startDate && event.timestamp < filter.startDate.getTime()) {
      return false;
    }
    
    if (filter.endDate && event.timestamp > filter.endDate.getTime()) {
      return false;
    }
    
    if (filter.severity && event.severity !== filter.severity) {
      return false;
    }
    
    if (filter.type && event.type !== filter.type) {
      return false;
    }
    
    if (filter.userId && event.metadata?.userId !== filter.userId) {
      return false;
    }
    
    return true;
  }
}

File: sdks/python/hypersim_sdk/security/security_auditor.py
"""
Security audit logging for Python SDK
"""

import asyncio
import json
import os
import time
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional

from .types import SecurityEvent, SecurityEventType


class SecurityAuditor:
    """Security audit logging and monitoring"""
    
    def __init__(
        self,
        enabled: bool,
        log_directory: Optional[str] = None,
        max_log_size: int = 10 * 1024 * 1024,  # 10MB
        rotate_daily: bool = True,
        enable_remote_logging: bool = False,
        remote_endpoint: Optional[str] = None,
        buffer_size: int = 100,
        flush_interval: int = 30
    ):
        self.enabled = enabled
        self.log_directory = log_directory or os.path.join(
            os.path.expanduser('~'), '.hypersim-audit'
        )
        self.max_log_size = max_log_size
        self.rotate_daily = rotate_daily
        self.enable_remote_logging = enable_remote_logging
        self.remote_endpoint = remote_endpoint
        self.buffer_size = buffer_size
        self.flush_interval = flush_interval
        
        self.event_buffer: List[SecurityEvent] = []
        self.current_log_file: Optional[str] = None
        
        if enabled:
            asyncio.create_task(self._initialize())
    
    async def log(self, event: SecurityEvent):
        """Log security event"""
        if not self.enabled:
            return
        
        # Add to buffer
        self.event_buffer.append(event)
        
        # Flush if buffer is full or event is critical
        if (len(self.event_buffer) >= self.buffer_size or 
            event.severity.value == 'critical'):
            await self.flush()
    
    async def log_auth(self, success: bool, user_id: Optional[str] = None, 
                      ip: Optional[str] = None, user_agent: Optional[str] = None):
        """Log authentication event"""
        event = SecurityEvent(
            event_type=SecurityEventType.UNAUTHORIZED_ACCESS,
            severity='low' if success else 'medium',
            timestamp=int(time.time() * 1000),
            description='Successful authentication' if success else 'Failed authentication attempt',
            metadata={
                'user_id': user_id,
                'ip': ip,
                'user_agent': user_agent,
                'success': success
            }
        )
        await self.log(event)
    
    async def log_api_usage(self, endpoint: str, method: str, status_code: int, 
                           response_time: float, user_id: Optional[str] = None):
        """Log API usage"""
        severity = 'medium' if status_code >= 400 else 'low'
        
        event = SecurityEvent(
            event_type=SecurityEventType.UNAUTHORIZED_ACCESS,
            severity=severity,
            timestamp=int(time.time() * 1000),
            description=f'API {method} {endpoint} - {status_code}',
            metadata={
                'endpoint': endpoint,
                'method': method,
                'status_code': status_code,
                'response_time': response_time,
                'user_id': user_id
            }
        )
        await self.log(event)
    
    async def log_data_access(self, resource: str, action: str, 
                             user_id: Optional[str] = None, success: bool = True):
        """Log data access"""
        event = SecurityEvent(
            event_type=SecurityEventType.UNAUTHORIZED_ACCESS,
            severity='low' if success else 'high',
            timestamp=int(time.time() * 1000),
            description=f'Data {action} on {resource}',
            metadata={
                'resource': resource,
                'action': action,
                'user_id': user_id,
                'success': success
            }
        )
        await self.log(event)
    
    async def flush(self):
        """Flush buffered events to storage"""
        if not self.event_buffer:
            return
        
        events = self.event_buffer.copy()
        self.event_buffer.clear()
        
        try:
            # Write to file
            await self._write_to_file(events)
            
            # Send to remote endpoint if enabled
            if self.enable_remote_logging and self.remote_endpoint:
                await self._send_to_remote(events)
        
        except Exception as error:
            # Add events back to buffer if write failed
            self.event_buffer.extend(events)
            print(f'Failed to flush audit logs: {error}')
    
    async def query_logs(self, filter_params: Dict) -> List[SecurityEvent]:
        """Query audit logs with filtering"""
        # In production: implement proper log querying
        return []
    
    async def get_stats(self, time_range: Dict) -> Dict:
        """Get security statistics"""
        # In production: implement statistics calculation
        return {
            'total_events': 0,
            'events_by_severity': {},
            'events_by_type': {},
            'top_ips': [],
            'top_users': []
        }
    
    async def archive_logs(self, older_than_days: int):
        """Archive old logs"""
        # In production: implement log archival
        pass
    
    async def shutdown(self):
        """Shutdown auditor"""
        await self.flush()
    
    async def _initialize(self):
        """Initialize audit logging"""
        if not self.enabled:
            return
        
        # Create log directory
        try:
            Path(self.log_directory).mkdir(parents=True, exist_ok=True, mode=0o700)
        except Exception as error:
            print(f'Failed to create audit log directory: {error}')
        
        # Set up periodic flushing
        asyncio.create_task(self._periodic_flush())
        
        self.current_log_file = self._get_current_log_file()
    
    async def _write_to_file(self, events: List[SecurityEvent]):
        """Write events to log file"""
        if not self.current_log_file:
            self.current_log_file = self._get_current_log_file()
        
        log_entries = []
        for event in events:
            log_entries.append(json.dumps({
                'type': event.event_type.value,
                'severity': event.severity.value,
                'timestamp': event.timestamp,
                'description': event.description,
                'metadata': event.metadata
            }))
        
        try:
            with open(self.current_log_file, 'a', encoding='utf-8') as f:
                f.write('\n'.join(log_entries) + '\n')
            
            # Check file size for rotation
            if os.path.getsize(self.current_log_file) > self.max_log_size:
                await self._rotate_log()
        
        except Exception as error:
            print(f'Failed to write audit log: {error}')
            raise
    
    async def _send_to_remote(self, events: List[SecurityEvent]):
        """Send events to remote endpoint"""
        # In production: implement remote logging with proper error handling
        pass
    
    async def _periodic_flush(self):
        """Periodic flush task"""
        while True:
            await asyncio.sleep(self.flush_interval)
            await self.flush()
    
    def _get_current_log_file(self) -> str:
        """Get current log file path"""
        today = datetime.now().strftime('%Y-%m-%d')
        return os.path.join(self.log_directory, f'security-audit-{today}.log')
    
    async def _rotate_log(self):
        """Rotate log file"""
        timestamp = datetime.now().strftime('%Y-%m-%d-%H-%M-%S')
        rotated_file = self.current_log_file.replace('.log', f'-{timestamp}.log')
        
        try:
            os.rename(self.current_log_file, rotated_file)
            self.current_log_file = self._get_current_log_file()
        except Exception as error:
            print(f'Failed to rotate log file: {error}')
